src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -0,0 +1,652 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace System.Formats.Tar
+{
+    // Writes header attributes of a tar archive entry.
+    internal partial struct TarHeader
+    {
+        private const byte SpaceChar = 0x20;
+        private const byte EqualsChar = 0x3d;
+        private const byte NewLineChar = 0xa;
+        private static readonly byte[] s_paxMagic = new byte[] { 0x75, 0x73, 0x74, 0x61, 0x72, 0x0 }; // "ustar\0"
+        private static readonly byte[] s_paxVersion = new byte[] { 0x30, 0x30 }; // "00"
+
+        private static readonly byte[] s_gnuMagic = new byte[] { 0x75, 0x73, 0x74, 0x61, 0x72, 0x20 }; // "ustar "
+        private static readonly byte[] s_gnuVersion = new byte[] { 0x20, 0x0 }; // " \0"
+
+        // Extended Attribute entries have a special format in the Name field:
+        // "{dirName}/PaxHeaders.{processId}/{fileName}{trailingSeparator}"
+        private const string PaxHeadersFormat = "{0}/PaxHeaders.{1}/{2}{3}";
+
+        // Global Extended Attribute entries have a special format in the Name field:
+        // "{tmpFolder}/GlobalHead.{processId}.1"
+        private const string GlobalHeadFormat = "{0}/GlobalHead.{1}.1";
+
+        // Creates a PAX Global Extended Attributes header and writes it into the specified archive stream.
+        internal static void WriteGlobalExtendedAttributesHeader(Stream archiveStream, IEnumerable<KeyValuePair<string, string>> globalExtendedAttributes)
+        {
+            TarHeader geaHeader = default;
+            geaHeader._name = GenerateGlobalExtendedAttributeName();
+            geaHeader._mode = (int)TarHelpers.DefaultMode;
+            geaHeader._typeFlag = TarEntryType.GlobalExtendedAttributes;
+            geaHeader._linkName = string.Empty;
+            geaHeader._magic = string.Empty;
+            geaHeader._version = string.Empty;
+            geaHeader._gName = string.Empty;
+            geaHeader._uName = string.Empty;
+            geaHeader.WriteAsPaxExtendedAttributes(archiveStream, globalExtendedAttributes, isGea: true);
+        }
+
+        // Writes the current header as a V7 entry into the archive stream.
+        internal void WriteAsV7(Stream archiveStream)
+        {
+            byte[] nameBytes = new byte[FieldLengths.Name];
+            byte[] modeBytes = new byte[FieldLengths.Mode];
+            byte[] uidBytes = new byte[FieldLengths.Uid];
+            byte[] gidBytes = new byte[FieldLengths.Gid];
+            byte[] sizeBytes = new byte[FieldLengths.Size];
+            byte[] mTimeBytes = new byte[FieldLengths.MTime];
+            byte[] checksumBytes = new byte[FieldLengths.Checksum];
+            byte typeFlagByte = 0;
+            byte[] linkNameBytes = new byte[FieldLengths.LinkName];
```

Comment:
Same question about allocation as earlier.  Seems like there should be more efficient ways to do this.

Reply:
Fixed with ArrayPool.